### PR TITLE
Refact replica clean up with supporting data-locality and eviction requested

### PR DESCRIPTION
**Pull request by boknowswiki**
1. Revert back the replica clean up for corrupted or staled case.
    2. Refactor replica clean up for failed to scheduled case.
    3. Add replica clean up support for data-locality enabled and eviction requested case.

https://github.com/longhorn/longhorn/issues/1679

Signed-off-by: Bo Tao <bo.tao@rancher.com>

### CODE REVIEW

File: controller/volume_controller.go [hunk 1]
```diff
@@ -604,9 +604,11 @@ func (vc *VolumeController) cleanupCorruptedOrStaleReplicas(v *longhorn.Volume,
 			// Skip running healthy replicas.
 			// Skip degraded volume.
 			// Skip if eviction is going.
+			// Skip if datalocality has been enabled
 			if r.Spec.HealthyAt != "" ||
 				healthyCount < v.Spec.NumberOfReplicas ||
-				hasEvictionRequestedReplicas {
+				hasEvictionRequestedReplicas ||
+				!isDataLocalityDisabled(v) {
```
Comment (yasker) [inline controller/volume_controller.go:611]: Should we limit to the replica that exists on the same attached node of v? cc @PhanLe1010

**Comment by shuo-wu** [comment]
Not sure why we need this PR? Can you clarify what happens (without this PR) if the data locality is enabled?

**Comment by PhanLe1010** [comment]
Right now, the PR only delete extra healthy replica when EvictionRequested or DataLocality are on. Should we also delete extra healthy replica in normal cases to keep a consistent behavior as discussed with @yasker and @boknowswiki yesterday?

**Comment by boknowswiki** [comment]
The document update has been submitted on the [same PR as 298 document PR](https://github.com/longhorn/website/pull/186).

**Comment by boknowswiki** [comment]
> Right now, the PR only deletes extra healthy replica when EvictionRequested or DataLocality are on. Should we also delete extra healthy replica in normal cases to keep a consistent behavior as discussed with @yasker and @boknowswiki yesterday?

I've thought about that as well. To make sure we are talking about the scenario.
User case with 3 worker nodes cluster, not regarding to eviction and data-locality.
1. Create a volume with 3 replicas. This will succeed.
2. User change the replica number to 2. In this case, there is one extra healthy replica.

Do you mean we should remove this extra replica? @PhanLe1010 

I think we should not remove it automatically. User may have plan with it, because we don't know which one user may want to delete or user may just want to leave it there. Since this extra replica is not driven by us, we should not touch it.

Please let me know if my understand is not correct or you have different ideas. Thanks.

**Comment by PhanLe1010** [comment]
@boknowswiki 
Yeah, I am referring about the above scenario. Because we are deleting extra healthy replicas when DataLocality and Eviction are on, I am thinking whether we should do the same in normal cases so that users have uniform expectation `Longhorn will always maintain exactly NumberOfReplicas replicas`. 

Having extra healthy replica laying around is not always a good thing. It creates more load on the network. So if the users decrease the NumberOfReplicas, it is more likely that they want to decrease the actual number of replicas.

>  User may have plan with it or user may just want to leave it there
If the users have a plan with the replica, I am not sure why they decrease the NumberOfReplicas.

>  we don't know which one user may want to delete 
Yeah, I agree with this one.

File: controller/volume_controller.go [hunk 2]
```diff
@@ -624,16 +623,89 @@ func (vc *VolumeController) cleanupCorruptedOrStaleReplicas(v *longhorn.Volume,
...
+
+	for _, r := range rs {
+		if r.Status.EvictionRequested == true {
+			hasEvictionRequestedReplicas = true
+		}
+	}
+
+	if healthyCount >= v.Spec.NumberOfReplicas {
+		for _, r := range rs {
+			if !hasEvictionRequestedReplicas {
```
Comment (yasker) [inline controller/volume_controller.go:650]: Hmm? Are we going to have the same issue of repeatedly deleting the local replica in this case?

**Review by yasker — CHANGES_REQUESTED** [review]

File: controller/volume_controller.go (hunk 2 above)
Comment (boknowswiki) [reply controller/volume_controller.go:650]: Hi Sheng,

Sorry, I don't fully understand your question. This is for eviction requested volume, in the case of there is no enough resource to evict the replica, we will have one failed to scheduled replica waiting for user ether to add more resource(node) or cancel the eviction. Thanks.

Comment (yasker) [reply controller/volume_controller.go:650]: I meant, are we going to keep the unscheduled replica due to data locality?

**Comment by PhanLe1010** [comment]
> Hmm? Are we going to have the same issue of repeatedly deleting the local replica in this case? I thought we're going to change the scheduling mechanism for the local replica to be the same as others, or are we still checking it separately?
> I meant, are we going to keep the unscheduled replica due to data locality?

we will discuss this over zoom.

**Review by yasker — APPROVED** [review]

**Comment by PhanLe1010** [comment]
Memory Note: In this PR, we don't delete extra healthy replicas when DataLocality and Eviction are off.

**Pull request merged**